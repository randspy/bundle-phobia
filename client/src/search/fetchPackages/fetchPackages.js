import fetchFrom from '../../shared/fetch-from/fetch-from';
import PackageBom from '../package-bom/package-bom';

export default async function fetchPackages(value) {
  if (hasAtLeastTwoCharacters(value)) {
    const response = await fetchFrom(
      `https://api.npms.io/v2/search/suggestions?q=${value}`
    );
    return response && Array.isArray(response)
      ? response.map(item => new PackageBom(item))
      : [];
  }
  return [];
}

function hasAtLeastTwoCharacters(value) {
  return value && value.length > 1;
}
