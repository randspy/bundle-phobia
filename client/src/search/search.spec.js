import React from 'react';
import { shallow } from 'enzyme';
import { Redirect } from 'react-router-dom';
import Search from './search';
import SelectPackage from './select-package/select-package';
import PackageBom from './package-bom/package-bom';
import fetchPackages from './fetchPackages/fetchPackages';

jest.mock('./fetchPackages/fetchPackages');

function mockPromise(value) {
  return fetchPackages.mockImplementation(() => {
    return Promise.resolve(value);
  });
}

describe('Search', () => {
  test('should wire component', () => {
    const wrapper = shallow(<Search />);
    expect(wrapper.exists()).toBe(true);
  });

  test('should contain SelectPackage', () => {
    const wrapper = shallow(<Search />);
    expect(wrapper.containsMatchingElement(<SelectPackage />)).toEqual(true);
  });

  test('should set options drop down list based on input', done => {
    const dataFromOnChange = { name: 'react' };
    const returnData = new PackageBom({ package: dataFromOnChange });
    const mockedFunction = mockPromise(returnData);
    const wrapper = shallow(<Search />);
    wrapper.find(SelectPackage).prop('onChange')(dataFromOnChange);

    setTimeout(() => {
      expect(wrapper.find(SelectPackage).props().options).toEqual(returnData);
      expect(mockedFunction).toHaveBeenCalledWith(dataFromOnChange);
      done();
    });
  });

  test('should redirect to result page when package is selected', () => {
    const dataFromOnChange = { name: 'react', version: '16.2.1' };
    const wrapper = shallow(<Search />);
    wrapper.find(SelectPackage).prop('onSelect')(dataFromOnChange);
    expect(wrapper.find(Redirect).props().to).toEqual(
      '/result?name=react&version=16.2.1'
    );
  });

  it('should render correctly', () => {
    const wrapper = shallow(<Search />);
    expect(wrapper).toMatchSnapshot();
  });
});
