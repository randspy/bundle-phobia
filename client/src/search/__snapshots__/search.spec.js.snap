// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should render correctly 1`] = `
<div
  className="search"
>
  <div>
    <div
      className="search--title"
    >
      BUNDLE
      <div
        className="search--secondary-title"
      >
        PHOBIA
      </div>
    </div>
    <div
      className="search--subtitle"
    >
      find the cost of adding a npm package to your bundle
    </div>
    <div>
      <SelectPackage
        onChange={[Function]}
        onSelect={[Function]}
        options={Array []}
      />
    </div>
  </div>
</div>
`;
